### ISSUE TITLE
[BUG] - Mock reliable collections do not serialize, or at least, deep copy inserted data

### ISSUE THREAD

**Issue by enioluwas** [loekd/ServiceFabric.Mocks#121]
**Describe the bug**

In the [reliable collections docs](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-work-with-reliable-collections#common-pitfalls-and-how-to-avoid-them), the authors describe scenarios in which programmers might accidentally cause bugs by updating objects in-memory, but not in the reliable collections. This is an important behavioral difference that reliable collections have over regular collections - updating a reference to data that was inserted into the reliable collection does not update the stored data. 

However, in our unit tests, we have noticed that the mock reliable collections do not model this behavior. Updating a reference to data that is stored in a mock reliable collection will apply to the stored data. This makes our unit tests behave differently than in integration tests, and has been the source of some confusion. Thus, in unit tests we have resorted to manually making deep copies of the data before storing them, by doing a serialization round trip.

Reliable collections by default support [data contract serialization](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer?view=net-6.0). It might be some work to support the default serialization, and not everyone is guaranteed to use it. However, custom serialization is also supported through the [IReliableStateManager.TryAddStateSerializer()](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.tryaddstateserializer?view=azure-dotnet) method. I noticed that the [MockReliableStateManager.TryAddStateSerializer()](https://github.com/loekd/ServiceFabric.Mocks/blob/d22a0b502d484e0398dad0687dc093c9d1cb8b29/src/ServiceFabric.Mocks/MockReliableStateManager.cs#L202) method is currently stubbed out. Is it feasible to implement this method to search for a serializer for a particular type before inserting data, so a deep copy is guaranteed for that type? Since we already use custom serialization, this would be enough to fix this for us.

I decided to open an issue first, to gauge interest in supporting this behavior before attempting a contribution.

**To Reproduce**

Here is a program that demonstrates this, modeled after the example in the [reliable collections docs](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-work-with-reliable-collections#common-pitfalls-and-how-to-avoid-them).

```c#
namespace MockReliableCollectionsBug
{
    using Microsoft.ServiceFabric.Data;
    using Microsoft.ServiceFabric.Data.Collections;
    using ServiceFabric.Mocks;
    using System;
    using System.Threading.Tasks;

    public static class MockReliableCollectionsBug
    {
        public class User
        {
            public string Name { get; set; }
            public DateTime LastLogin { get; set; }
        }

        public static async Task Main(string[] args)
        {
            MockReliableStateManager stateManager = new MockReliableStateManager();
            IReliableDictionary2<string, User> userDict = await stateManager.GetOrAddAsync<IReliableDictionary2<string, User>>("test");
            User user = new User { Name = "Gail", LastLogin = DateTime.UtcNow };

            using (ITransaction tx = stateManager.CreateTransaction())
            {
                await userDict.AddAsync(tx, user.Name, user);
                await tx.CommitAsync();
            }

            // Update the in-memory user's LastLogin.
            user.LastLogin = DateTime.UtcNow;

            using (ITransaction tx = stateManager.CreateTransaction())
            {
                ConditionalValue<User> storedUser = await userDict.TryGetValueAsync(tx, user.Name);
                // This should output false, as the storedUser's LastLogin should be the original value, but
                // it returns true, since the stored value is a direct reference, not a copy of the in-memory value
                // which has since been updated.
                Console.WriteLine(user.LastLogin == storedUser.Value.LastLogin);
                await tx.CommitAsync();
            }
        }
    }
}
```

**Expected behavior**
Console should output `false`.

**Comment by loekd** [comment]
Hi, thanks for your feedback!

SF [guidance ](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-work-with-reliable-collections#define-immutable-data-types-to-prevent-programmer-error) explains that the state objects should be immutable. Following this guidance should mitigate your issue.

So, would this really be a good idea? I feel it would enable developers to do it the incorrect way..

**Comment by enioluwas** [comment]
> I feel it would enable developers to do it the incorrect way..

Interesting viewpoint... I think the opposite though - let me explain. The _real_ reliable collections already enable developers to do this. Using immutable types is a guidance, not a requirement. I think that the mocks _allowing_ direct reliable state mutation is actually part of what misleads developers believe they can do it in a real SF cluster. If unit tests with mocks could catch bugs due to "mistaken mutation", as described in the docs and in this issue, it would be less confusing for developers what can and can't be done with references to reliable state. Does that make sense?

**Comment by loekd** [comment]
Yep, that does sound nice. Let's discuss a bit how we can add this feature. 

It will need to be opt-in behavior, so it won't affect existing tests and performance in cases where the 'old' behavior is not influencing results. So no serialization happens if you don't call `TryAddStateSerializer.`

**Comment by enioluwas** [comment]
Agreed! Opt-in would be per-type, inherently, since `TryAddStateSerializer<T>` is for type `T`. 

How I imagine it would work is, the `MockReliableStateManager` would keep track of the registered serializers via some thread-safe type cache, and when constructing a mock collection, if either the `TKey` or `TValue` type has a serializer, it will be passed to the collection. Before collection writes are done, the values to be inserted can be deep copied by doing a round trip of `IStateSerializer.Write<T>` and `IStateSerializer.Read<T>`, if one exists for `T`. Then the write will proceed as normal. 

Does the general idea sound good?

**Comment by loekd** [comment]
Sounds like a plan. Let's make sure we follow the docs, concerning types (value types vs reference types).
I've also got the plan to deprecate .NET 4.5 support and move everything to .NET6. 
You can find the branch [here](https://github.com/loekd/ServiceFabric.Mocks/tree/CleanupAndSerialize).

**Comment by enioluwas** [comment]
I've made the changes on a fork based on the branch you linked to, how do you want the PR published, to merge into master or to merge into the linked branch?

**Comment by loekd** [comment]
Oh, I've also been busy with this. Can you share a link to your repo /branch so we can compare notes and create the best solution?
I've pushed mine to the branch.

**Comment by enioluwas** [comment]
> edit: I think it's [this fork](https://github.com/enioluwas/ServiceFabric.Mocks/commit/1eda3c853a146090d9b1fc4db20dce733adf1e5d) right?

Yep!

I just glanced through your commit and you're right, we had a pretty similar approach! I did make the reliable collection constructor with the serializers `internal`, and I see you made yours `public`. What do you prefer?

> What are your thoughts about distinguishing key serializers from value serializers?

For the reliable dictionary, the `TKey` and `TValue` will almost always be different types. I store the serializers for the key and value types in the constructor to avoid doing a lookup in the `ConcurrentDictionary` every time `MaybeCloneKey`/`MaybeCloneValue` happens (even in the case when there's no serializer registered). It would also enable some strange behavior, since the dictionary is mutable, and typically calls `TryAddStateSerializer` should not affect already recovered reliable collections. 

> And what's missing in both is dealing with inherited data contract types. I'm not sure how that works in SF. Let's say your TestUser is inherited as class TestAdmin : TestUser. Would that require a new serializer?

I don't think the [DataContract](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractattribute?view=net-6.0) attribute is inheritable. I think it has to be explicitly defined for each type:

> A data contract differs significantly from an actual interface in its inheritance behavior. Interfaces are inherited by any derived types. When you apply the [DataContractAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractattribute?view=net-6.0) to a base class, the derived types do not inherit the attribute or the behavior. However, if a derived type has a data contract, the data members of the base class are serialized. However, you must apply the [DataMemberAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datamemberattribute?view=net-6.0) to new members in a derived class to make them serializable.

So with that I would say yes, each type needs its own serializer. The way the type matching works [in SF](https://github.com/microsoft/service-fabric/blob/master/src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/SerializationManager/SerializationManager.cs) seems pretty similar actually.

**Comment by loekd** [comment]
I thought having the serializer collection argument visible would allow people to write unit tests without requiring a statemanager.
Ok, let's go with a 1:1 match between type and serializer for now.
I'll have another look at your separate keys & values serializers approach.

**Comment by enioluwas** [comment]
> I thought having the serializer collection argument visible would allow people to write unit tests without requiring a statemanager.

Ok that's fair. I didn't like exposing `ConcurrentDictionary<Type, object>` because we lose the internal guarantee that `object` will be `IStateSerializer<T>`, and without documentation it might not be immediately obvious what the constraints are on the dictionary to pass in.

> Oh btw, another difference is that I store serialized objects as streams internally, while you clone immediately.

You changed the backing store to store the serialized data? I actually thought about doing that, but it seemed like shifting the default behavior towards needing a serializer. In the case where none are registered, what would be the conversion from object to stream?

I decided on cloning because it seemed like the most non-invasive way of achieving the objective of not having direct references to reliable objects, either by reading or writing

**Comment by loekd** [comment]
I was going for an approach that would closely mimic the real deal (which uses file streams). The end result is the same I suppose.
Good point about the serializer type check.

I've also copied your key /value serializer approach in my latest push.

**Comment by loekd** [comment]
Would you mind reviewing [the PR](https://github.com/loekd/ServiceFabric.Mocks/pull/122)?
I think I managed to merge our solutions.

As a future feature we might be able to create a generic stateserializer using system.text.json?

**Comment by enioluwas** [comment]
Closing the issue was an accident, whoops. Sure, I'll review this later today, glad we could get this done! :smile: 

Hmm I think if it's supposed to emulate SF, then we would need a generic stateserializer that handles built-in types as well as `DataContract` serializable types. The SF implementations for [built-in](https://github.com/microsoft/service-fabric/blob/master/src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/SerializationManager/BuiltInTypeSerializer.cs) and [data contract](https://github.com/microsoft/service-fabric/blob/master/src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/SerializationManager/DataContractStateSerializer.cs) are public, so it should be possible! If we use System.Text.Json, then people have to write a serializer for that anyway, if they want to use it on the real SF cluster.

**Comment by enioluwas** [comment]
I'm starting to think some of my assumptions about reliable collections might be wrong. I just read [this doc](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-reliable-collections-guidelines) which says:
> - Do not modify an object of custom type returned by read operations (for example, TryPeekAsync or TryGetValueAsync). Reliable Collections, just like Concurrent Collections, return a reference to the objects and not a copy. 
> - Do deep copy the returned object of a custom type before modifying it. Since structs and built-in types are pass-by-value, you do not need to do a deep copy on them unless they contain reference-typed fields or properties that you intend to modify.

And then I re-read this part from [the doc](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-work-with-reliable-collections#common-pitfalls-and-how-to-avoid-them):

> When working with a regular .NET dictionary, you can add a key/value to the dictionary and then change the value of a property (such as LastLogin). However, this code will not work correctly with a reliable dictionary. Remember from the earlier discussion, the call to AddAsync serializes the key/value objects to byte arrays and then saves the arrays to a local file and also sends them to the secondary replicas. If you later change a property, this changes the property's value in memory only; it does not impact the local file or the data sent to the replicas. If the process crashes, what's in memory is thrown away. When a new process starts or if another replica becomes primary, then the old property value is what is available.

Basically, as long as the replica stays up, any modifications made to objects in reliable stores will be reflected in reads on that replica alone, but as soon as there is primary swap or a move, the modified object is lost, because it isn't serialized and replicated across the secondary replicas.

This is different and more complicated than what I believed the behavior was. I think the current implementation is right, and I can see even more the advantage in using immutable data structures here. I'm going to close this issue, given this new information. It was fun to figure out how to implement this change, unfortunate that it turned out not to be what was needed. Thanks for all the time you put in here! If you disagree with these findings, you can reopen the issue to discuss.
